File: algorithm/valid_anagram.c
#include <stdio.h>
#include <stdbool.h>

/*
 * https://leetcode.com/problems/valid-anagram/
 *
 * Time Limit Exceeded ...
 */

/* select sort */
void
sort(char *s)
{
    int i, max, index;
    char tmp;
    char *sp1 = s, *sp2;

    index = 0;
    while ('\0' != *sp1) {
        sp2 = sp1;
        i = index;
        max = index;
        while ('\0' != *sp2) {
            if (*sp2 > s[max]) {
                max = i;
            }
            ++i;
            ++sp2;
        }
        tmp = s[index];
        s[index] = s[max];
        s[max] = tmp;
        ++index;
        ++sp1;
    }
}

bool
is_anagram(char *s, char *t)
{
    sort(s);
    sort(t);

    while ('\0' != *s && '\0' != *t && *s == *t) {
        ++s;
        ++t;
    }

    return '\0' == *s && '\0' == *t ? true : false;
}

int
main()
{
    char s[] = "anagram";
    char t[] = "nagaram";
    char s1[] = "acac";
    char t1[] = "accc";
    char s2[] = "rat";
    char t2[] = "cat";

    printf("%d\n", is_anagram(s, t));
    printf("%d\n", is_anagram(s1, t1));
    printf("%d\n", is_anagram(s2, t2));

    return 0;
}

File: algorithms/valid_anagram.py
# -*- coding: utf-8 -*-

# https://leetcode.com/problems/valid-anagram/


def is_anagram(s, t):
    if len(s) != len(t):
        return False

    hts = dict()
    for c in s:
        hts[c] = hts.get(c, 0) + 1

    htt = dict()
    for c in t:
        htt[c] = htt.get(c, 0) + 1

    for k in hts:
        if k not in htt or htt[k] != hts[k]:
            return False

    return True


if __name__ == '__main__':
    print(is_anagram('anagram', 'nagaram'))
    print(is_anagram('rat', 'car'))
    print(is_anagram('ccac', 'acac'))
